### PR TITLE
Add two new metric cards for sessions.

### PR DESCRIPTION
**Pull request by krievley**
This pull request adds two new metric cards: Sessions and Avg. Session Duration. 

False Data is included in the screenshot below to show the format and style of the cards.

Screen Shot 2021-06-15 at 7 11 07 PM

### CODE REVIEW

File: src/SessionDurationMetric.php [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+use Laravel\Nova\Metrics\Value;
+use Laravel\Nova\Metrics\ValueResult;
+use Spatie\Analytics\Analytics;
+use Spatie\Analytics\Period;
+
+class SessionDurationMetric extends Value
+{
+    public function name() {
+        return __('Avg. Session Duration');
+    }
```
Comment (mattstauffer) [inline src/SessionDurationMetric.php]:
```suggestion
    public function name()
    {
        return __('Avg. Session Duration');
    }
```

File: src/SessionDurationMetric.php [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    /**
+     * Get the URI key for the metric.
+     *
+     * @return string
+     */
+    public function uriKey()
+    {
+        return 'session-duration';
+    }
+}
```
Comment (mattstauffer) [inline src/SessionDurationMetric.php]: Missing ending line

File: src/SessionDurationMetric.php [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+            [
+                'metrics' => 'ga:avgSessionDuration',
+                'dimensions' => 'ga:year',
+            ]
+        );
+
+        $previousResults = collect($previousAnalyticsData->getRows());
+
+        return [
+            'previous' => $previousResults->last()[1] ?? 0,
```
Comment (mattstauffer) [inline src/SessionDurationMetric.php:119]: Is this `??` available in our minimum PHP version? I assume so, but just wanted to make sure

File: src/SessionDurationMetric.php [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+            'result' => $currentResults->last()[1] ?? 0,
+        ];
+    }
+
+    private function sessionDurationOneYear()
+    {
+        $currentAnalyticsData = app(Analytics::class)->performQuery(
+            Period::create(Carbon::today()->startOfYear(), Carbon::today()),
```
Comment (mattstauffer) [inline src/SessionDurationMetric.php:96]: Is this correct for year? or should it be today sub year? It's always hard for me to know how that should be

Comment (mattstauffer) [reply src/SessionDurationMetric.php:96]: OK, so it looks like you're doing this year YTD compared to last year all year.

Should it instead be "the year that started one year ago and ends today, vs the year that started two years ago and ended one year ago"? I don't know what we're doing elsewhere

File: src/SessionsMetric.php [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+use Illuminate\Support\Arr;
+use Laravel\Nova\Metrics\Value;
+use Spatie\Analytics\Analytics;
+use Spatie\Analytics\Period;
+
+class SessionsMetric extends Value
+{
+    public function name() {
+        return __('Sessions');
+    }
```
Comment (mattstauffer) [inline src/SessionsMetric.php]:
```suggestion
    public function name()
    {
        return __('Sessions');
    }
```

File: src/SessionDurationMetric.php [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+        return [
+            'result' => $results->last()[1] ?? 0,
+            'previous' => $results->first()[1] ?? 0,
+        ];
+    }
+
+    private function sessionDurationOneMonth()
+    {
+        $currentAnalyticsData = app(Analytics::class)->performQuery(
+            Period::create(Carbon::today()->startOfMonth(), Carbon::today()),
```
Comment (mattstauffer) [inline src/SessionDurationMetric.php:65]: Same notes/questions here, and also on the sessions metric, that I gave on the year metric for this one.

**Review by mattstauffer — CHANGES_REQUESTED** [review]
Some questions here about how to handle dates

File: src/SessionDurationMetric.php (hunk 3 above)
Comment (krievley) [reply src/SessionDurationMetric.php:119]: Yes, we've utilized it for our Page Views metric with no issues.

File: src/SessionDurationMetric.php (hunk 4 above)
Comment (krievley) [reply src/SessionDurationMetric.php:96]: We are doing Today, MTD, and YTD on the metric cards.

Comment (mattstauffer) [reply src/SessionDurationMetric.php:96]: OK, that answers the first half of the question.

But it opens a new question (yay!) lol:

If we are doing "YTD" on july 1, part of me thinks the user wants to compare
"This year january 1 to july 1" vs "last year january 1 to july 1"
But this code right here shows

"this year january 1 to july 1" vs "last year january 1 to december 31"

So... which do users expect?

**Review by mattstauffer — CHANGES_REQUESTED** [review]
Great work! Let's see how the rest of the team thinks about the YTD functionality, but here are some notes on the rest.

File: src/MetricDiffTrait.php [hunk 7]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+namespace Tightenco\NovaGoogleAnalytics;
+
+use Carbon\Carbon;
+
+trait MetricDiffTrait
+{
+    private function getPeriodDiff($startDate) {
```
Comment (mattstauffer) [inline src/MetricDiffTrait.php:9]:
```suggestion
    private function getPeriodDiff($startDate)
    {
```

Comment (mattstauffer) [reply src/MetricDiffTrait.php:9]: Let's typehint startDate just for documentation

Comment (mattstauffer) [reply src/MetricDiffTrait.php:9]: I think this means:

Assuming today is the end of a period
and the provided startdate was the first day of that period
give me the start and end of the *previous* period

method name probably more like "getPreviousPeriodDates"

Maybe, in order to make this clearer:

```php
    private function getPreviousPeriodDates($thisPeriodStart, $thisPeriodEnd = null)
    {
        if (! $thisPeriodEnd) {
            $thisPeriodEnd = Carbon::today();
        }
        
        //
        
        return [$lastPeriodStart, $lastPeriodEnd];
    }
```

potential simpler code:

```php
$thisPeriodLength = $thisPeriodStart->diffInDays($thisPeriodEnd);
$lastPeriodEnd = $thisPeriodStart->clone()->subDay();
$lastPeriodStart = $lastPeriodEnd->clone()->subDays($thisPeriodLength);
```

**Review by mattstauffer — CHANGES_REQUESTED** [review]

**Pull request closed without merging**